source/Cosmos.Core/Memory/Heap.cs

```diff
@@ -32,6 +32,40 @@ public static unsafe void Init()
             HeapLarge.Init();
         }
 
+        /// <summary>
+		/// Re-allocates or "re-sizes" data asigned to a pointer.
+		/// </summary>
+		/// <param name="aPtr">Existing pointer</param>
+		/// <param name="NewSize">Size to extend to</param>
+		/// <returns>New pointer with specified size while maintaining old data.</returns>
+        public static byte* Realloc(byte* aPtr, uint NewSize)
+		{
+            // Get existing size
+            uint Size = (RAT.GetPageType(aPtr) == RAT.PageType.HeapSmall ? ((ushort*)aPtr)[-2] : ((uint*)aPtr)[-4]); 
+
+            if (Size == NewSize)
+			{
+                // Return existing pointer as nothing needs to be done.
+                return aPtr;
+			}
+            if (Size > NewSize)
+			{
+                Size -= (NewSize - Size);
+			}
+
+            // Allocate a new buffer to use
+			byte* ToReturn = Alloc(NewSize);
+
+            // Copy the old buffer to the new one
+            MemoryOperations.Copy(ToReturn, aPtr, (int)Size);
+            // Copy the GC state
+            ((ushort*)ToReturn)[-1] = ((ushort*)aPtr)[-1];
```

Comment:
I have thought about this more and it seems like we should just set the GC state to 0 for the new object, since the old references no longer count.